Stylish devdocs (#478)
OffchainLabs/arbitrum-docs: arbitrum-docs/stylus/stylus-gentle-introduction.md

> ---

Author: @amarrazza your gentle intro is here :)

title: 'A gentle introduction to Stylus'
sidebar_label: 'A gentle introduction to Stylus'
description: ''
author: amarrazza
sme: amarrazza
target_audience: todo
sidebar_position: 1
---

# A gentle introduction: Stylus

import PublicPreviewBannerPartial from './partials/_stylus-public-preview-banner-partial.md';

<PublicPreviewBannerPartial />

> This introduction is for developers who are interested in learning about Stylus, a new way to write smart contracts in popular programming languages, like Rust, alongside the EVM.

Author: ```suggestion
This introduction is for developers who want to build Ethereum decentralized apps (dApps) using Rust. This capability is made possible by **Stylus**, a new way to write EVM-compatible smart contracts using your favorite programming languages.
```

A suggestion, feel free to discard, nits welcome!

Rationale -> we want to open with the _**keywords that devs use to search for solutions to their problem**_ first, _then_ our solution to that problem.

Reviewer: I am wary about saying "Ethereum" dapps because it implies the contracts are on L1 Ethereum and only for app builders. To me, the quickstart is specifically targeted at "developers who want to build Ethereum decentralized apps (dApps) using Rust" where this is more educational. I think only including Rust for the audience could be limiting too, since Stylus is about other languages and contributing libraries for non-supported languages as well. I do support giving Rust the spotlight though! 

How about this?

"This introduction is for developers who want to build on Arbitrum using popular programming languages, like Rust. This capability is made possible by **Stylus**, a new way to write EVM-compatible smart contracts using your favorite programming languages.

Reviewer: I pushed a commit with the compromise, but happy to continue riffing on this

### In a nutshell:

- Stylus lets you write smart contracts in programming languages that compile down to WASM, such as **Rust, C, C++, and many others**.

> - Rich language support already exists for Rust: developers can use the Stylus SDK and CLI tool to **start building today.**

Author: ```suggestion
- Rich language support already exists for Rust: you can use the Stylus SDK and CLI tool to **start building today.**
```

Reviewer: I like it

> - Arbitrum's full **EVM equivalence would be maintained** while adding a second, coequal WASM virtual machine.

Author: ```suggestion
- Stylus smart contracts benefit from **Arbitrum's full EVM equivalence** thanks to a second, coequal WASM virtual machine.
```

Reviewer: I love it

> - Due to the superior efficiency of WASM programs, Stylus contracts are over **an order of magnitude faster with significantly lower gas fees**.

Author: ```suggestion
- Stylus contracts are over **an order of magnitude faster with significantly lower gas fees** due to the superior efficiency of WASM programs.
```

(value first, explanation after)

> - **Memory is 100-500x cheaper** in Stylus, enabling new use cases now that consuming RAM is viable on the blockchain.

Author: ```suggestion
- **Memory is 100-500x cheaper** when using Stylus; this unlocks new use-cases that were previously too expensive to implement on-chain.
```

definitely a nit, feel free to discard any of these suggestions that don't strike you as "pure wins"

Reviewer: consuming RAM gets at a specific pain point solidity devs experience, so I would vote to keep that part in

With a WASM VM, any programming language that can compile down to WASM are within the bounds of Stylus. While many popular programming languages can be compiled into WASM, some compilers are more suitable for smart contract development than others, like with Rust, C, and C++. Other languages, such as Go, Sway, Move, and Cairo, can be supported as well. Languages that include their own runtimes, like Python and Javascript, are harder for Stylus to support, although not impossible. Third-party contribution in the form of libraries for new and existing languages is welcomed!

Compared to using Solidity, WASM programs are much more efficient. There are many reasons for this, including the decades of compiler development for Rust and C. WASM also has a faster runtime than the EVM, resulting in faster execution. Generally, a **10x improvement** has been seen for contracts using WASM languages compared to contracts using Solidity.

### How is this possible?

> Stylus is only possible because of Arbitrum Nitro's unique fraud-proving technology. When there is a dispute on an Arbitrum network, Nitro replays the execution of the chain **in WASM.** Honest Arbitrum validators will then bisect what is being disputed until a single invalid step is identified and checked on-chain via a [“one-step proof.”](https://docs.arbitrum.io/proving/challenge-manager#general-bisection-protocol)

Author: ```suggestion
Stylus is only possible because of Arbitrum Nitro's unique fraud-proving technology. When there's a dispute on an Arbitrum network, Nitro replays the execution of the chain **in WASM.** Honest Arbitrum validators will then bisect what is being disputed until a single invalid step is identified and checked on-chain via a [“one-step proof.”](https://docs.arbitrum.io/proving/challenge-manager#general-bisection-protocol)
```

voice nit -> friendly contractions as a way to "write like we're speaking to a friend"

#### Fully composable

Solidity programs and WASM programs are completely interoperable. If working in Solidity, a developer can call a Rust program or rely on another dependency in a different language. If working in Rust, all Solidity functionalities are accessible out of the box.

### How does it work?

There are four main steps for bringing a Stylus program to life — coding, compilation, execution, and proving.

#### Coding

In principle, developers can write smart contracts in any programming language that can be compiled into WASM. In practice, some high-level languages generate far more performant WASMs than others.

Initially, there will be support for Rust, C, and C++. However, the levels of support will differ at first. Rust has rich language support from day one, with an open-source SDK that makes writing smart contracts in Rust as easy as possible. C and C++ are supported off the bat too, which will enable deploying existing contracts in those languages on-chain with minimal modifications.

> The Stylus SDK for Rust contains the smart contract development framework and language features most developers will need to use Stylus. The SDK also makes it possible to perform all of the EVM-specific functionalities that smart contract developers are used to. Check out the official [Stylus SDK](https://docs.arbitrum.io/) and [Rust crate](https://docs.arbitrum.io/).

Author: ```suggestion
The Stylus SDK for Rust contains the smart contract development framework and language features most developers will need to use Stylus. The SDK also makes it possible to perform all of the EVM-specific functionalities that smart contract developers are used to. Check out the official [Stylus SDK](https://docs.arbitrum.io/) and [Rust crate](https://docs.rs/crate/stylus-sdk).
```

todo - SDK doc

Reviewer: I'm gonna update all links to docs now
